### ISSUE TITLE
AltTab sometimes doesn't respond quickly; it has high CPU usage

### ISSUE THREAD

**Issue by yarmoliq** [lwouis/alt-tab-macos#3986]
**Describe the bug**

After recent update there is a big delay between I can switch between windows and see any pop up (basically using alttab). Looking at activity monitor, I see that alttab creates huge spikes of CPU usage. Yesterday when this started happening I updated again, but it didn't fix the issue. I have reported this bug through built-in feedback thingy with cpu information. Just wanted to duplicate it here so there is awareness.

<img width="818" alt="image" src="https://github.com/user-attachments/assets/a3a998ec-e437-4a7a-aa5e-d9120a48cfc1" />

**Your environment**

* AltTab version: 7.12.0
* macOS version: 15.2 (24C101)
* Other relevant info: no nothing

<details>
<summary><strong>Debug profile</strong></summary>
<p>

* App version: 7.12.0
* App preferences: 
  * MSAppCenter310AppCenterUserDefaultsMigratedKey: 1
  * MSAppCenter310CrashesUserDefaultsMigratedKey: 1
  * MSAppCenterInstallId: BEC9E98A-0C10-4934-96AB-9B26316D53E3
  * MSAppCenterNetworkRequestsAllowed: 0
  * MSAppCenterPastDevices: {length = 1144, bytes = 0x62706c69 73743030 d4010203 04050607 ... 00000000 00000384 }
  * MSAppCenterSessionIdHistory: {length = 460, bytes = 0x62706c69 73743030 d4010203 04050607 ... 00000000 00000156 }
  * MSAppCenterUserIdHistory: {length = 455, bytes = 0x62706c69 73743030 d4010203 04050607 ... 00000000 00000151 }
  * NSStatusItem Visible Item-0: 0
  * NSWindow Frame SUStatusFrame: 828 862 400 135 0 0 2056 1285 
  * NSWindow Frame SUUpdateAlert: 718 665 620 398 0 0 2056 1285 
  * SUHasLaunchedBefore: 1
  * SULastCheckTime: 2024-12-15 10:38:15 +0000
  * SUUpdateGroupIdentifier: 2750686940
  * SUUpdateRelaunchingMarker: 0
  * alignThumbnails: 0
  * appearanceSize: 0
  * appearanceStyle: 1
  * appsToShow: 0
  * fontHeight: 14
  * hideAppBadges: true
  * hideSpaceNumberLabels: true
  * hideStatusIcons: true
  * hideThumbnails: true
  * hideWindowlessApps: true
  * holdShortcut: ⌘
  * iconSize: 40
  * menubarIcon: 0
  * menubarIconShown: false
  * preferencesVersion: 7.12.0
  * previousWindowShortcut: ⇧⇥
  * screenRecordingPermissionSkipped: true
  * screensToShow: 0
  * showMinimizedWindows: 0
  * showOnScreen: 1
  * showTabsAsWindows: false
  * spacesToShow: 1
  * theme: 0
  * titleTruncation: 2
  * updatePolicy: 1
  * windowMaxWidthInRow: 7
  * windowMinWidthInRow: 7
* Applications: 63
* Windows: 13
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: true, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: true, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: false, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: false, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: true, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: true, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: true, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: true, isHidden: false, isTabbed: false, isOnAllSpaces: true, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: false, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: true, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: true, isHidden: false, isTabbed: false, isOnAllSpaces: true, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: false, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
  * {isFullscreen: false, isWindowlessApp: false, isMinimized: false, isHidden: false, isTabbed: false, isOnAllSpaces: false, shouldShowTheUser: true, spaceId: 3, spaceIndex: 3}
* OS version: Version 15.2 (Build 24C101)
* OS architecture: arm64
* Locale: en_US (fixed en_US)
* Spaces: 4
* Dark mode: Dark
* "Displays have separate Spaces": unchecked
* Hardware model: Mac14,6
* Screens: 1
  * {x: 0.0, y: 0.0, width: 2056.0, height: 1329.0}
* CPU model: Apple M2 Max
* Memory size: 34.36 GB
* Active CPU count: 12
* Current CPU frequency: nil
* Resource utilization: 
  * CPU: 0.5%
  * Memory: 185M+
  * Threads count: 13

</p>
</details>

**Comment by lwouis** [comment]
Hi,
Thank you for sharing this issue.
Could you please follow instructions from this message?
Thank you 🙇

**Comment by timothyvermeiren** [comment]
Hi All! Visiting from #3845, hopefully with relevant debug logs. I think was able to capture this as it happened, when I pressed CMD+tab at `11:41:05`.

Attached are the logs from that timeframe:
* A few lines from before.
* The lines from when the freeze occurred, notably with a gap between `11:41:05` stating `handleKeyboardEvent()` and `11:41:11`.
* I think that like most people I pressed CMD+tab a number of times as a reaction to AltTab not showing, which probably explains the rather large number of `handleKeyboardEvent()` repetitions all at `11:41:14` which is probably when the app un-froze.

Happy to provide more details if I can.

[AltTab-debug.log](https://github.com/user-attachments/files/18148750/AltTab-debug.log)

**Comment by lwouis** [comment]
Thank you @timothyvermeiren 

It's hard to follow the logs without knowing exactly what you did at the same time, and what you saw as well.

You mentioned a lack of logs between `11:41:05` and `11:41:11`. At `11:41:05`, we see the option key down, I think. Without `tab` being pressed. At `11:41:11`, we see alt+tab are down, thus it triggers. It looks like normal behavior here. I don't know what you pressed though.

In the logs, we see a lot of `showUiOrCycleSelection()` -> `focusTarget()` -> `hideUi()`. Looking at the code, it seems that you may have ended up in `checkForFingersUp()` which triggers `focusTarget` when the 3 fingers are released from the trackpad. Do you use the Trackpad (Preferences > Controls > Gesture tab)?

Thank you

**Comment by lwouis** [comment]
For everyone else: please share a video with the logs. It will help a lot understand the sequence, and read the logs.

**Comment by yarmoliq** [comment]
I just tried using alttab normally. It takes a while for it to unfreez if I don't switch windows for a minute or 2. Once I press it, there is no response for 2-10 seconds. Then a window finally switches. It doesn't matter if I press command tab once or if I press it continuously until it responds. Also, once it does respond, it doesn't matter how fast I press command tab, cpu utilization doesn't rise at all. It only happens when it is frozen.

> Do you use the Trackpad (Preferences > Controls > Gesture tab)?

I'm not sure what is that. I don't use trackpad to switch between windows at all. Occasionally to switch between virtual desktops

**Comment by timothyvermeiren** [comment]
> Thank you @timothyvermeiren
> 
> It's hard to follow the logs without knowing exactly what you did at the same time, and what you saw as well.
> 
> You mentioned a lack of logs between `11:41:05` and `11:41:11`. At `11:41:05`, we see the option key down, I think. Without `tab` being pressed. At `11:41:11`, we see alt+tab are down, thus it triggers. It looks like normal behavior here. I don't know what you pressed though.
> 
> In the logs, we see a lot of `showUiOrCycleSelection()` -> `focusTarget()` -> `hideUi()`. Looking at the code, it seems that you may have ended up in `checkForFingersUp()` which triggers `focusTarget` when the 3 fingers are released from the trackpad. Do you use the Trackpad (Preferences > Controls > Gesture tab)?
> 
> Thank you

Hello again, @lwouis.

I don't have a video that goes with that specific log entry, but I can tell you what happened. I checked the time, to the second, before I pressed the shortcut that I have assigned to AltTab: CMD + tab. It was 11:41:05, without a doubt.

That's why I am noting the lack of logs specifically. I realize the logs make it seem like a different shortcut was pressed; possibly the option key when my finger slipped on it?

As for the trackpad gestures: I don't think I am using that. In fact, it doesn't seem to be available on the version (7.8.0) I have now:

![CleanShot 2024-12-17 at 06 09 48](https://github.com/user-attachments/assets/c0a58b1c-d8af-4497-975b-aba355f921be)

Regarding the video: I can't promise anything. I realize it would be easier, but it's tough to record your (very large) screen all the time while working on several things, having client calls, etc.

**Comment by yarmoliq** [comment]
maybe it has something to do with the 15.2 macos update?

**Comment by deepaktalwardt** [comment]
I'm having the exact same issue. AltTab spikes up CPU usage, holds it up for up to 10 seconds at a time. During this time I experience lags with scrolling etc; and Cmd+Tab doesn't show the AltTab switcher UI. It then hangs and window only switches after a few seconds. 
 
I'm on MacOS 15.2 as well - although I have experienced this on and off in earlier MacOS versions as well. It seems to have gotten worse now. 

My guess is that it might have something to do with the screen capture requirement for getting previews, but I could be wrong.

<img width="954" alt="image" src="https://github.com/user-attachments/assets/c23c3aea-2c8d-49f3-8adf-f5c100c55766" />

**Comment by Duc-NSH** [comment]
I do not have the high CPU usage issue, but it is really delayed when I try to use the mouse to hover too fast among windows, the delay can be up to 2 seconds

**Comment by Duc-NSH** [comment]
tried to revert the version to 7.9.0, but the issue remains, I doubt that there the MacOS 15.2 did something to alt-tab

**Comment by morhekil** [comment]
I'm seeing the same problem on 14.6.1. AltTab sometimes doesn't respond to keystrokes, and its menu bar icon just shows a spinner when hovered. It unfreezes by itself after 10 seconds or so. I started seeing this only recently, so seems to be related to one of the more recent AltTab releases.

**Comment by lwouis** [comment]
Thank you everyone for sharing your story.

If we want progress on this, we need data. As it is, I have nothing to investigate.

if you can, please share a video of the issue. Also please share a spindump, as well as the logs.

Thank you

**Comment by yarmoliq** [comment]
how do i make a spindump? and where do i find the logs?
also, i have previews turn off, so it probably has nothing to do with that

**Comment by yarmoliq** [comment]
i will do that. where will the logs be?

**Comment by lwouis** [comment]
In Terminal.app, right bellow the command you'll enter

**Comment by deepaktalwardt** [comment]
> Could you please share a spindump when the app is high CPU and unresponsive? You can generate one via Activity Monitor. Thank you

I just sampled the AltTab process as it spun up to high CPU usage again. What's the best way to share it?

**Comment by lwouis** [comment]
Please share the file here directly. You can attach it to your comment

**Comment by deepaktalwardt** [comment]
> Please share the file here directly. You can attach it to your comment

[Sample of AltTab.txt](https://github.com/user-attachments/files/18187713/Sample.of.AltTab.txt)

**Comment by elliottower** [comment]
I created #4008 through the app, sorry didn't check the issues. Commenting here in case it's fixed. Pretty bad experience right now, may have to use stock apple alt tab behavior until it's fixed

**Comment by morhekil** [comment]
just as another point of reference – I downgraded AltTab to 7.0.0, and it seems that the problem is gone now, I'm not getting any frequent freezes anymore

**Comment by deepaktalwardt** [comment]
Another thing I noticed today -  It seems to only happen when I'm using the MacBook's display. I had been using my Mac in docked mode with the external display for the last couple of days and I didn't notice any slowdowns. Today, however, within the first 5 minutes of opening the Mac without the display connected, I noticed AltTab spin up and slow down. I have also noticed WindowServer spin up along side AltTab - maybe that's related? 

I also noticed that WindowServer spun up to 100% usage a few moments after I killed AltTab. Hopefully these help!

**Comment by lwouis** [comment]
Hi everyone,

I've done a large rework. Could you please try out this [local build](https://github.com/user-attachments/files/18288899/AltTab.zip), and share how it performs?

Thank you

**Comment by lwouis** [comment]
Hi everyone,
The latest build contains performance improvements.
Could you please test it and let me know if the issue in this ticket still occurs?
Thank you!

**Comment by lucypero** [comment]
I've been using the new version for a while today and it hasn't frozen so far :)

**Comment by morhekil** [comment]
new build is working well for me, too – I didn't get any CPU spinners so far with it

**Comment by timothyvermeiren** [comment]
Still the same behavior here, unfortunately. In fact, it's sometimes at a point where AltTab stops responding and instead of "unfreezing" after ~20 seconds, it remains stuck. Not sure if that indicates anything relevant.

I wasn't running in debug mode unfortunately (I will do that now for a while) but I did catch
 a spindump from the event. [Spindump 2025-01-06T08-46.txt.zip](https://github.com/user-attachments/files/18315351/Spindump.2025-01-06T08-46.txt.zip)

**Comment by lwouis** [comment]
Hi,
I've just released a new version. I think it should fix this issue.
Could you please try it out, and let me know?
Thank you! 🙇

**Comment by timothyvermeiren** [comment]
Thanks you for remaining so responsive, @lwouis, and for the quick turnaround with which you're addressing what comes up. Really appreciate it.

I'm going to install 7.18 (that one right) and give it a whirl for a few days, then report back here!

**Comment by cokert** [comment]
I was seeing hangs very regularly (multiple times an hour) until updating to 7.18.0 yesterday but it's been solid since the update.  🤞

**Comment by lwouis** [comment]
The issues seem gone. I'll close this ticket. If you still face issues, please let me know 👍

**Comment by zacharee** [comment]
I'm not sure what triggered it, but the lag is back for me in 7.18.0 and 7.18.1. It was fine for a while and then suddenly this morning it started taking up to 30 seconds to respond to the keyboard shortcut, the settings window shows the spinning beachball a lot, and CPU usage is spiking.

**Comment by lwouis** [comment]
Hi @zacharee,
Could you please share a spindump when it happens again?
Thank you
